src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java

```diff
@@ -458,13 +473,15 @@ private void handleProxyConnectionError(Throwable error) {
      * Called when proxy failed to receive response header from origin.
      */
     private void handleProxyResponseError(Throwable error) {
+        UpstreamRoute upstreamRoute = context.getUpstreamRoute();
         log.warn("Proxy failed to receive response header from origin. Trace: {}. Span: {}. Key: {}. Deployment: {}. Address: {}. Error:",
                 context.getTraceId(), context.getSpanId(),
                 context.getProject(), context.getDeployment().getName(),
                 context.getProxyRequest().connection().remoteAddress(),
                 error);
 
-        context.getUpstreamRoute().retry();
+        upstreamRoute.fail(HttpStatus.BAD_GATEWAY, -1);
```

Comment:
see javadoc description

Reply:
Perhaps it’s worth adding one more method in this case.
Since that confuses

Reply:
fail5xx fail429? I think this may lead to more complex error handling
We can add a comment in all places where we pass -1

Reply:
I'm for clear methods where arguments make sense. You pass -1 just to pass something. That's not good. You could make an arg as optional like in Python but Java doesn't support that. I would overload the methods that's more simple to me.

Reply:
my main point - call single method in case of any error. Method itself responsible for the logic how to dial with specific errors.
with your approach you push every developers (who will use this class) to handle this logic by themselves and check every time what's the type of error

We can introduce 2 methods, but they must follow single logic, for example
```
public synchronized void failed(HttpStatus status, long retryAfterSeconds) {
        if (status == HttpStatus.TOO_MANY_REQUESTS) {
            retryAfter = System.currentTimeMillis() + retryAfterSeconds * 1000;
        }
                if (status.is5xx()) {
            if (++errorCount >= errorsThreshold) {
                retryAfter = System.currentTimeMillis() + (long) (INITIAL_BACKOFF_DELAY_MS * Math.pow(BACKOFF_BASE, errorCount));
            }
        }
    }

public synchronized void failed(HttpStatus status) {
failed(status, 30);
    }
    ```